datadog_checks_base/datadog_checks/base/checks/windows/perf_counters/base.py

```diff
@@ -47,19 +55,35 @@ def query_counters(self):
         with self.adopt_namespace(self.namespace):
             self._query_counters()
 
+    def should_refresh_enum_obj_list(self):
+        now = time.time()
+        return now - self.last_enum_obj_list_refresh > self.enum_object_list_refresh_duration
+
     def _query_counters(self):
         # Refresh the list of performance objects, see:
         # https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhenumobjectitemsa#remarks
-        try:
-            # https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhenumobjectsa
-            # https://mhammond.github.io/pywin32/win32pdh__EnumObjects_meth.html
-            win32pdh.EnumObjects(None, self._connection.server, win32pdh.PERF_DETAIL_WIZARD, True)
-        except pywintypes.error as error:
-            message = 'Error refreshing performance objects: {}'.format(error.strerror)
-            self.submit_health_check(self.CRITICAL, message=message)
-            self.log.error(message)
 
-            return
+        if self.should_refresh_enum_obj_list():
+            with self.enum_object_list_lock.write_lock():
```

Comment:
Was the idea here to serialize enumerations, or to "deduplicate" enumerations when multiple requests for enumeration arrive at a time it needs to be refreshed?  I ask because this will serialize, but not deduplicate -- the duplicate calls will just "queue up" and enumerate one after the next.

Reply:
@djmitche excellent catch!!! Yes, idea was deduplicate. Maybe double-checked locking patter is suited here better https://en.wikipedia.org/wiki/Double-checked_locking

Reply:
@iglendd I ended up using a nonblocking lock since that seems like a simpler solution.